File: samples/cpp/tutorial_code/Histograms_Matching/calcBackProject_Demo2.cpp
/**
 * @file BackProject_Demo2.cpp
 * @brief Sample code for backproject function usage ( a bit more elaborated )
 * @author OpenCV team
 */

#include "opencv2/imgproc.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui.hpp"

#include <iostream>

using namespace cv;
using namespace std;

/// Global Variables
Mat src, hsv, mask;

int low = 20, up = 20;
const char* window_image = "Source image";

/// Function Headers
void Hist_and_Backproj( );
void pickPoint (int event, int x, int y, int, void* );

/**
 * @function main
 */
int main( int, char** argv )
{
    /// Read the image
    src = imread( argv[1] );

    /// Transform it to HSV
    cvtColor( src, hsv, COLOR_BGR2HSV );

    /// Show the image
    namedWindow( window_image );
    imshow( window_image, src );

    /// Set Trackbars for floodfill thresholds
    createTrackbar( "Low thresh", window_image, &low, 255, 0 );
    createTrackbar( "High thresh", window_image, &up, 255, 0 );
    /// Set a Mouse Callback
    setMouseCallback( window_image, pickPoint, 0 );

    waitKey();
    return 0;
}

/**
 * @function pickPoint
 */
void pickPoint (int event, int x, int y, int, void* )
{
    if( event != EVENT_LBUTTONDOWN )
    {
        return;
    }

    // Fill and get the mask
    Point seed = Point( x, y );

    int newMaskVal = 255;
    Scalar newVal = Scalar( 120, 120, 120 );

    int connectivity = 8;
    int flags = connectivity + (newMaskVal << 8 ) + FLOODFILL_FIXED_RANGE + FLOODFILL_MASK_ONLY;

    Mat mask2 = Mat::zeros( src.rows + 2, src.cols + 2, CV_8U );
    floodFill( src, mask2, seed, newVal, 0, Scalar( low, low, low ), Scalar( up, up, up), flags );
    mask = mask2( Range( 1, mask2.rows - 1 ), Range( 1, mask2.cols - 1 ) );

    imshow( "Mask", mask );

    Hist_and_Backproj( );
}

/**
 * @function Hist_and_Backproj
 */
void Hist_and_Backproj( )
{
    Mat hist;
    int h_bins = 30; int s_bins = 32;
    int histSize[] = { h_bins, s_bins };

    float h_range[] = { 0, 180 };
    float s_range[] = { 0, 256 };
    const float* ranges[] = { h_range, s_range };

    int channels[] = { 0, 1 };

    /// Get the Histogram and normalize it
    calcHist( &hsv, 1, channels, mask, hist, 2, histSize, ranges, true, false );

    normalize( hist, hist, 0, 255, NORM_MINMAX, -1, Mat() );

    /// Get Backprojection
    Mat backproj;
    calcBackProject( &hsv, 1, channels, hist, backproj, ranges, 1, true );

    /// Draw the backproj
    imshow( "BackProj", backproj );
}

File: samples/python/tutorial_code/Histograms_Matching/back_projection/calcBackProject_Demo2.py
from __future__ import print_function
import cv2 as cv
import numpy as np
import argparse

low = 20
up = 20

def callback_low(val):
    global low
    low = val

def callback_up(val):
    global up
    up = val

def pickPoint(event, x, y, flags, param):
    if event != cv.EVENT_LBUTTONDOWN:
        return

    # Fill and get the mask
    seed = (x, y)
    newMaskVal = 255
    newVal = (120, 120, 120)
    connectivity = 8
    flags = connectivity + (newMaskVal << 8 ) + cv.FLOODFILL_FIXED_RANGE + cv.FLOODFILL_MASK_ONLY

    mask2 = np.zeros((src.shape[0] + 2, src.shape[1] + 2), dtype=np.uint8)
    print('low:', low, 'up:', up)
    cv.floodFill(src, mask2, seed, newVal, (low, low, low), (up, up, up), flags)
    mask = mask2[1:-1,1:-1]

    cv.imshow('Mask', mask)
    Hist_and_Backproj(mask)

def Hist_and_Backproj(mask):
    h_bins = 30
    s_bins = 32
    histSize = [h_bins, s_bins]
    h_range = [0, 180]
    s_range = [0, 256]
    ranges = h_range + s_range # Concat list
    channels = [0, 1]

    # Get the Histogram and normalize it
    hist = cv.calcHist([hsv], channels, mask, histSize, ranges, accumulate=False)
    cv.normalize(hist, hist, alpha=0, beta=255, norm_type=cv.NORM_MINMAX)

    # Get Backprojection
    backproj = cv.calcBackProject([hsv], channels, hist, ranges, scale=1)

    # Draw the backproj
    cv.imshow('BackProj', backproj)

# Read the image
parser = argparse.ArgumentParser(description='Code for Back Projection tutorial.')
parser.add_argument('--input', help='Path to input image.')
args = parser.parse_args()

src = cv.imread(args.input)
if src is None:
    print('Could not open or find the image:', args.input)
    exit(0)

# Transform it to HSV
hsv = cv.cvtColor(src, cv.COLOR_BGR2HSV)

# Show the image
window_image = 'Source image'
cv.namedWindow(window_image)
cv.imshow(window_image, src)

# Set Trackbars for floodfill thresholds
cv.createTrackbar('Low thresh', window_image, low, 255, callback_low)
cv.createTrackbar('High thresh', window_image, up, 255, callback_up)
# Set a Mouse Callback
cv.setMouseCallback(window_image, pickPoint)

cv.waitKey()
